Sources/ASTGen/ASTGen.swift

```diff
@@ -0,0 +1,253 @@
+import SwiftParser
+import SwiftSyntax
+
+import CASTBridging
+
+extension Array {
+  public func withBridgedArrayRef<T>(_ c: (BridgedArrayRef) -> T) -> T {
+    withUnsafeBytes { buf in
+      c(BridgedArrayRef(data: buf.baseAddress!, numElements: count))
+    }
+  }
+}
+
+extension UnsafePointer {
+  public var raw: UnsafeMutableRawPointer {
+    UnsafeMutableRawPointer(mutating: self)
+  }
+}
+
+struct ASTGenVisitor: SyntaxTransformVisitor {
+  let ctx: UnsafeMutableRawPointer
+  let base: UnsafePointer<CChar>
+
+  // TOOD: we need to be up updating this.
+  var declContext: UnsafeMutableRawPointer
+  
+  // TODO: this some how messes up the witness table when I uncomment it locally :/
+//  public func visit<T>(_ node: T?) -> [UnsafeMutableRawPointer]? {
+//    if let node = node { return visit(node) }
+//    return nil
+//  }
+  
+  @_disfavoredOverload
+  public func visit(_ node: SourceFileSyntax) -> UnsafeMutableRawPointer {
+    fatalError("Use other overload.")
+  }
+  
+  public func visit(_ node: SourceFileSyntax) -> [UnsafeMutableRawPointer] {
+    let loc = self.base.advanced(by: node.position.utf8Offset).raw
+    var out = [UnsafeMutableRawPointer]()
+
+    for element in node.statements {
+      let swiftASTNodes = visit(element)
+      if element.item.is(StmtSyntax.self) {
```

Comment:
I think we should encode information that elements in a source file can only be `StmtSyntax`, `ExprSyntax` or `DeclSyntax` in the syntax tree so we can do an exhaustive switch here.

The best idea that I’ve come up with is to code-gen enums like the following based on the `node_choices` of `CodeBlockItem.Item` in [CommonNodes.py](https://github.com/apple/swift-syntax/blob/main/Sources/generate-swift-syntax-builder/gyb_syntax_support/CommonNodes.py#L37-L43)
```swift
public struct CodeBlockItemSyntax /* CodeBlockItemSyntax conformances */ {
  enum Item {
    case stmt(StmtSyntax)
    case expr(ExprSyntax)
    case decl(DeclSyntax)
    case tokenList(TokenListSyntax)
    case nonEmptyTokenList(NonEmptyTokenListSyntax)
  
    func as(_ t: StmtSyntax.Type) -> StmtSyntax? { … }
    func as<T: StmtSyntaxProtocol>(_ t: T.Type) -> T? { … }
    func as(_ t: ExprSyntax.Type) -> ExprSyntax? { … }
    func as<T: ExprSyntaxProtocol>(_ t: T.Type) -> T? { … }
    func as(_ t: DeclSyntax.Type) -> DeclSyntax? { … }
    func as<T: DeclSyntaxProtocol>(_ t: T.Type) -> T? { … }
    func as(_ t: TokenList.Type) -> TokenListSyntax? { … }
    func as(_ t: nonEmptyTokenList.Type) -> NonEmptyTokenListSyntax? { … }
  }
  // existing code in CodeBlockItemSyntax
}
```

and change the type of `item` in CodeBlockItemSyntax from `Syntax` to `CodeBlockItemSyntax.Item`.
I think (haven’t tested) that this should be mostly (i.e. covering most common API use cases) source-comaptible with the current usage of these APIs.

That way you could exhaustivly switch over it here as
```swift
switch element.item {
    case stmt(StmtSyntax):
      out.append(SwiftTopLevelCodeDecl_createStmt(ctx, declContext, loc, swiftASTNodes, loc))
    case expr(ExprSyntax):
      out.append(SwiftTopLevelCodeDecl_createExpr(ctx, declContext, loc, swiftASTNodes, loc))
    case decl(DeclSyntax):
      out.append(swiftASTNodes)
    case tokenList(TokenListSyntax):
      fatalError("Should not be generated by SwiftParser")
    case nonEmptyTokenList(NonEmptyTokenListSyntax):
      fatalError("Should not be generated by SwiftParser")
}
```

I would really like to see this refactoring done before we fully start ASTGen. I don’t think it needs to block this PR but it’s something we should do ASAP.

A quick improvement without this refactoring: You can use `element.item.as(SyntaxEnum.self)` here and then switch over the enum cases.

Reply:
I think this would be absolutely fantastic! The hardest thing about working with the syntax tree is that you don't know what cases to cover at any given point without working back from the grammar.

Reply:
All that said, I don't know that *this* pull request is the place for this discussion. It's a general improvement to the ergonomics of the SwiftSyntax APIs.